service-base/src/main/java/org/eclipse/hono/service/registration/BaseRegistrationService.java

```diff
@@ -304,6 +322,43 @@ public void assertRegistration(
         }).setHandler(resultHandler);
     }
 
+    private Future<Void> updateDeviceLastViaIfNeeded(final String tenantId, final String deviceId,
+            final String gatewayId, final JsonObject deviceData) {
+        if (!isDeviceWithMultipleVias(deviceData)) {
+            return Future.succeededFuture();
+        }
+        return updateDeviceLastVia(tenantId, deviceId, gatewayId, deviceData);
+    }
+
+    private boolean isDeviceWithMultipleVias(final JsonObject deviceData) {
+        final Object viaObj = deviceData.getValue(PROPERTY_VIA);
+        return viaObj instanceof JsonArray && ((JsonArray) viaObj).size() > 1;
+    }
+
+    /**
+     * Updates device registration data and adds a 'last-via' property containing the given gateway identifier
+     * as well as the current date.
+     * <p>
+     * This method this called by this class' default implementation of <em>assertRegistration</em> for a device that
+     * has multiple gateway entries in its 'via' definition.
+     * <p>
+     * This default implementation simply returns a failed Future with a {@link ServerErrorException}
+     * with status code 501 (Not Implemented).
+     * Subclasses need to override this method and provide a reasonable implementation in order for this
+     * class' default implementation of <em>assertRegistration</em> to work properly when using devices
+     * that have multiple gateway entries in their 'via' definition.
+     *
+     * @param tenantId The tenant id.
+     * @param deviceId The device id.
+     * @param gatewayId The gateway id.
+     * @param deviceData The current data associated with the device.
+     * @return A future indicating whether the operation succeeded or not.
+     */
+    protected Future<Void> updateDeviceLastVia(final String tenantId, final String deviceId, final String gatewayId,
+            final JsonObject deviceData) {
+        return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_NOT_IMPLEMENTED));
```

Comment:
I wonder if the default implementation should not return a succeeded future instead. The authorization check has already been performed so this method basically represents *optional* behavior. If a subclass wants to support it, then it needs to override it with an appropriate implementation. However, if not, then nothing would need to be done.
On the other hand, if we want to force people to think about this feature when creating a subclass, then we should define this method as abstract. WDYT?

Reply:
I would rather make the method abstract then, describing in the javadoc that a proper implementation is needed for getting command & control to work in the scenario described in #930 (ie. device with multiple `via` gateways; gateways subscribing to commands without specifying the device id).
If such a scenario is not important for someone implementing the device registry service, the method can then be implemented by just returning a succeeded future.

Reply:
I've amended the commit accordingly.